### PR TITLE
Caching with key func

### PR DESCRIPTION
**Pull request by koreno**

### CODE REVIEW

File: easypy/caching.py [hunk 1]
```diff
@@ -171,29 +172,73 @@ def clear():
...
+                elif param.kind is _VAR_POSITIONAL:
+                    val = ()
+                elif param.kind is _VAR_KEYWORD:
+                    val = {}
+                else:
+                    # This BoundArguments was likely produced by
+                    # Signature.bind_partial().
+                    continue
+                new_arguments.append((name, val))
+        bound_arguments.arguments = OrderedDict(new_arguments)
```
Comment (idanarye) [inline easypy/caching.py]: Why convert a list of pairs to an `OrderedDict`? Why not just make an empty `OrderedDict` and populate it?

File: easypy/caching.py [hunk 2]
```diff
@@ -171,29 +172,73 @@ def clear():
...
+else:
+    def _apply_defaults(bound_arguments):
+        return bound_arguments.apply_defaults()
```
Comment (idanarye) [inline easypy/caching.py]: `bound_arguments.apply_defaults()` returns `None`, but it's still weird that this version "returns" something while the backward-compatibility version doesn't.

File: easypy/caching.py (hunk 1 above)
Comment (koreno) [reply easypy/caching.py]: I copied the code from python3.5 sources, didn't want to change anything to official code.

File: easypy/caching.py (hunk 2 above)
Comment (koreno) [reply easypy/caching.py]: I'm like a decorator, don't care about implementation detail, I want to be future-compatible.

Comment (idanarye) [reply easypy/caching.py]: Future-compatibility-wise, if in the future apply_default will return something, someone might use the value returned from _apply_defaults believing it's backward compatible

File: tests/test_misc.py [hunk 3]
```diff
@@ -7,3 +7,14 @@ def foo(p=AUTO):
...
+    assert MAX == "MAX"
+    assert MAX == "<MAX>"
+    assert MAX == "max"
+    assert MAX == "<max>"
+
+    d = {AUTO: AUTO, MAX: MAX}
+    assert d[AUTO] == AUTO
+    assert d[MAX] == MAX
+    assert d['<MAX>'] is MAX
+    assert 'AUTO' not in d
```
[the remark this answers is not in the document]
Comment (koreno) [reply tests/test_misc.py]: I'll revisit this at some point...

**Comment by koreno** [comment]
(already merged)

**Pull request closed without merging**